src/lib.rs

```diff
@@ -0,0 +1,534 @@
+        crate::unix::memory_resize_and_lock(memory, allow_mem_resize)
+    }
+}
+
+fn memory_unlock(memory: &mut [usize]) -> anyhow::Result<()> {
+    #[cfg(windows)]
+    {
+        crate::windows::memory_unlock(memory)
+    }
+
+    #[cfg(unix)]
+    {
+        crate::unix::memory_unlock(memory)
+    }
+}
+
+#[cfg(windows)]
+mod windows {
+    use {
+        crate::prelude::*,
+        std::mem::MaybeUninit,
+        windows::Win32::{
+            Foundation::ERROR_WORKING_SET_QUOTA,
+            System::{
+                Memory::{VirtualLock, VirtualUnlock},
+                SystemInformation::{GetNativeSystemInfo, SYSTEM_INFO},
+                Threading::{
+                    GetCurrentProcess, GetProcessWorkingSetSize, SetProcessWorkingSetSize,
+                },
+            },
+        },
+    };
+
+    // TODO: We don't really need to replace the set size if if it is much larger than memsize?
+    pub(super) fn replace_set_size(memsize: usize) -> anyhow::Result<(usize, usize)> {
+        let (mut min_set_size, mut max_set_size) = (0, 0);
+        unsafe {
+            GetProcessWorkingSetSize(GetCurrentProcess(), &mut min_set_size, &mut max_set_size)
+                .context("Failed to get process working set")?;
+            // TODO: Not sure what the best choice of min and max should be
+            SetProcessWorkingSetSize(
+                GetCurrentProcess(),
+                memsize.saturating_mul(2),
+                memsize.saturating_mul(4),
+            )
+            .context("Failed to set process working set")?;
+        }
+        Ok((min_set_size, max_set_size))
+    }
+
+    pub(super) fn restore_set_size(min_set_size: usize, max_set_size: usize) -> anyhow::Result<()> {
+        unsafe {
+            SetProcessWorkingSetSize(GetCurrentProcess(), min_set_size, max_set_size)
+                .context("Failed to set process working set")
+        }
+    }
+
+    // TODO: Resize according to min set size instead of decrementing
+    pub(super) fn memory_resize_and_lock<'a>(
+        mut memory: &'a mut [usize],
+        allow_mem_resize: bool,
+    ) -> anyhow::Result<&'a mut [usize]> {
+        let page_size = unsafe {
+            let mut sysinfo: MaybeUninit<SYSTEM_INFO> = MaybeUninit::uninit();
+            GetNativeSystemInfo(sysinfo.as_mut_ptr());
+            let sysinfo = sysinfo.assume_init();
+            usize::try_from(sysinfo.dwPageSize).context("Failed to convert page size to usize")?
+        };
+
+        loop {
+            unsafe {
+                match VirtualLock(memory.as_mut_ptr().cast(), size_of_val(memory)) {
+                    Ok(()) => {
+                        info!("Successfully locked {}MB", size_of_val(memory));
+                        return Ok(memory);
+                    }
+                    Err(e) if e.code() == ERROR_WORKING_SET_QUOTA.to_hresult() => {
+                        if allow_mem_resize {
+                            match size_of_val(memory).checked_sub(page_size) {
```

Comment:
I think it would simplify things and get rid of a bunch of "byte len <=> usize len" conversions if you add something like this right before the loop:

```Rust
let usize_per_page = page_size / std::mem::size_of::<usize>();
```

Then, you can replace this with:

```Rust
let new_len = memory
    .len()
    .checked_sub(usize_per_page)
    .context("failed to lock any memory, memory size has been decremented to 0")?;
memory = &mut memory[0..new_len];
```